grpc/swift-cubic/Sources/swift-cubic/CubicManager.swift

```diff
@@ -0,0 +1,168 @@
+        self.client = Cobaltspeech_Cubic_CubicServiceClient(connection: connection)
+    }
+    
+    public init(host: String,
+                port: Int,
+                tlsCertificateFileName: String,
+                tlsCertificateFormat: NIOSSLSerializationFormats) {
+        let target = ConnectionTarget.hostAndPort(host, port)
+        let eventLoopGroup = PlatformSupport.makeEventLoopGroup(loopCount: 1,
+                                                                networkPreference: .best)
+        
+        var configuration = ClientConnection.Configuration(target: target, eventLoopGroup: eventLoopGroup)
+        
+        if let cert = try? NIOSSLCertificate(file: tlsCertificateFileName, format: tlsCertificateFormat) {
+            let source = NIOSSLCertificateSource.certificate(cert)
+            var tls = ClientConnection.Configuration.TLS()
+            tls.certificateChain.append(source)
+            configuration.tls = tls
+        }
+        
+        let connection = ClientConnection.init(configuration: configuration)
+        self.client = Cobaltspeech_Cubic_CubicServiceClient(connection: connection)
+    }
+    
+    public func listModels(success: @escaping (_ models: [Cobaltspeech_Cubic_Model]?) -> (), failure: CubicFailureCallback?) {
+        let request = Cobaltspeech_Cubic_ListModelsRequest()
+        
+        client.listModels(request).response.whenComplete({ (result) in
+            do {
+                let response = try result.get()
+                success(response.models)
+            } catch let error {
+                failure?(error)
+            }
+        })
+    }
+    
+    public func sendStreamPartToServer(_ audioData: Data, failure: CubicFailureCallback?) {
+        guard let callStream = callStream else { return }
+        
+        var request = Cobaltspeech_Cubic_StreamingRecognizeRequest()
+        request.audio.data = audioData
+        
+        callStream.sendMessage(request).whenComplete({ (result) in
+            switch result {
+            case .failure(let error):
+                failure?(error)
+            default:
+                break
+            }
+        })
+    }
+    
+    public func startStream(success: @escaping (_ response: Cobaltspeech_Cubic_RecognitionResponse) -> (), failure: CubicFailureCallback?) {
+        guard let selectedModel = selectedModel else { return }
+        
+        let call = self.client.streamingRecognize(handler: { (result) in
+            success(result)
+        })
+
+        var request = Cobaltspeech_Cubic_StreamingRecognizeRequest()
+        request.config.modelID = selectedModel.id
+
+        if let idleTimeoutSeconds = idleTimeoutSeconds {
+            request.config.idleTimeout.seconds = idleTimeoutSeconds
+        }
+        
+        if let audioEncoding = audioEncoding {
+            request.config.audioEncoding = audioEncoding
+        }
+
+        call.sendMessage(request).whenComplete({ (result) in
+            switch result {
```

Comment:
where is the result processed/displayed?

Reply:
6 lines above. There is a handler for streamRecognize which receives Cobaltspeech_Cubic_RecognitionResponse.

And here we just send a message with config without audio.

Reply:
can you give example code for a customer to call startStream by providing their own function to process the responses?

Sorry, this is mostly because of my limited experience with swift; so I'm unable to connect the dots. Documentation should help understand this.

Reply:
I finally understood your question. )) Would it be better if I unite three methods (`startStream`, `sendStreamPartToServer` and `stopStream`) into one `streamRecognize`, like it is done in other languages examples?